docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,341 @@
+| `tx`                 | * hash <br/> * prove                   | Get a transaction by its hash                   | This companion endpoint will return the same response structure as the CometBFT endpoint.                                                                                                                                                                                               |
+| `validators`         | * height <br/> * page <br/> * per_page | Get validator set at a specified height         | This companion endpoint will return the same response structure as the CometBFT endpoint.                                                                                                                                                                                               |
+
+The following endpoints can also be implemented, but might require some additional effort and complexity to be implemented. These are mostly the `search` and `query` endpoints.
+
+| **JSONRPC method**   | **JSONRPC Parameters**                                               | **Description**                        | **Notes**                                                                                                                                                                                    |
+|----------------------|----------------------------------------------------------------------|----------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `abci_query`         | * path <br/> * data <br/> * height <br/> * prove                     | Query information from the application | This companion endpoint will return the same response structure as the CometBFT endpoint. The RPC companion service will have to implement a proper abci parameter to sql query translation  |
+| `block_search`       | * query <br/> * page <br/> * per_page <br/> * order_by               | Query information about a block        | This companion endpoint will return the same response structure as the CometBFT endpoint. The RPC companion service will have to implement a proper query parameter to sql query translation |
+| `tx_search`          | * query <br/> * page <br/> * per_page <br/> * prove <br/> * order_by | Query information about transactions   | This companion endpoint will return the same response structure as the CometBFT endpoint. The RPC companion service will have to implement a proper query parameter to sql query translation |
```

Comment:
According to my understanding:

* offloading `block_search` and `tx_search` methods to the companion has the biggest potential of improving full node's performance
* other methods whose implementation can simply access the block store or state store directly are interesting from a storage perspective, but less impactful performance-wise

So would be very beneficial to have `block_search` and `tx_search` in the RPC companion. The problem I see for including them in an MVP (maybe that's why they are in this list?) is that, ideally, we would use the underlying database's SQL querying, rather than migrating our ad-hoc indexing code. If so, we couldn't offer _the same_ querying language to the RPC-companion's users as full nodes implement.

Is this discussion worth including in this ADR?

Reply:
This comment is along the lines with Jasimna comments and I think this is probably a totally different discussion (or even an ADR). There are still many unknowns to validate before we can attest that this idea will work. I agree and think that the "search" and "query" would be the most beneficial but requires some effort to get it to work (if feasible). So I'll just leave the additional conversation to the near future but not part of this ADR